### ISSUE TITLE
Selecting outermost core nodes for drainage

### ISSUE THREAD

**Issue by amanaster2** [landlab/landlab#1888]
Hi all!

I have a bit of a weird question, and I wasn't finding the answer to it in the documentation:
TL;DR Is there a way to select the outermost core nodes on a raster and determine the # of nodes draining to them? (That isn't just brute force?)

**My problem:**
I'm working with DEMs of gravel forest road surfaces to see how micro-topographical changes affect water flow pathways. I have water flowing off the sides of the road as well as down the road, and I want to determine the # of nodes draining to each of the nodes along the sides and bottom of my road surface. However, my DEMs are not perfectly rectangular, so I can't simply select the nodes along the grid edges. Is there a way to do this in Landlab? And, if so, will I be able to (easily) delineate between the sides and bottom?

**Example surface water discharge map:**
![wsd_example](https://github.com/landlab/landlab/assets/26778806/0ae6af4d-378a-4f2a-8f47-99518dd92169)

Hope you're doing well!
Amanda

**Comment by gregtucker** [comment]
Hi @amanaster2 nice to hear from you! Have you tried querying all boundary nodes with a drainage area or discharge greater than zero? Something like:

```
area = grid.at_node["drainage_area"]
is_drainage_boundary = np.logical_and(area > 0, grid.status_at_node != grid.BC_NODE_IS_CORE)
boundary_drainage_area = area[is_drainage_boundary]
```

**Comment by gregtucker** [comment]
By default, the perimeter nodes in a RasterModelGrid will be set to BC_NODE_IS_FIXED_VALUE status, and all the others are set to core. From the imagine above, I'm guessing you probably overrode this to set all the nodes with an elevation of -9999 to some kind of boundary status (fixed value or closed)? Maybe using `set_watershed_boundary_conditions()`? If they are set to fixed value status, and you run FlowAccumulator, it will record a drainage area for each fixed-value boundary node that receives flow from an adjacent core node.

Quick example:

```
grid = RasterModelGrid((4, 6))
z = grid.add_zeros('topographic__elevation', at='node')
z[:] = np.arange(grid.number_of_nodes)
fa = FlowAccumulator(grid)
fa.run_one_step()
grid.at_node["drainage_area"].reshape((4, 6))
```
Should display:
```
array([[ 0.,  2.,  2.,  2.,  2.,  0.],
       [ 0.,  2.,  2.,  2.,  2.,  0.],
       [ 0.,  1.,  1.,  1.,  1.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.]])
```
The fact that the value 2 appears in the top row means that those boundary nodes received drainage from the adjacent core nodes (with a drainage area of 2 m2 each).

**Comment by amanaster2** [comment]
I did override the BC_NODE_IS_FIXED_VALUE status by using `set_nodata_nodes_to_closed()`. I actually hacked my output a little bit, and I'm thinking that I maybe made a mess of things by doing so... I'm going to try my best to explain this.

When I set my nodata nodes to closed **_before_** running FlowAccumulator, I end up with water being routed along the edge of my road segment (which is not what physically happens because the water should continue to be diverted off of the sides).
Example (sorry, shifting sites): 
![image](https://github.com/landlab/landlab/assets/26778806/e138ac6e-334e-4b9d-bce7-4c1c11a7f8af)

 So instead, I'm running FlowAccumulator, THEN setting my nodata nodes to closed for a more realistic drainage path.
 Example:
![image](https://github.com/landlab/landlab/assets/26778806/3a3d148f-a7d7-4994-acb6-74eccd049371)

By doing this, though, I think I'm grabbing incorrect data if I use the code you first suggested.

It's almost like I need to be able to use `set_watershed_boundary_conditions()`, but with my right, left, and bottom road edges ALL being allowed to be outlets. I'm trying to get a sense of where water is exiting the system and how the average location shifts over time.

**Comment by gregtucker** [comment]
How do you envision the terrain adjacent to your road? For example, is there a steep (higher) bank on one side? If so, that seems like a good case for a closed boundary (unless water is seeping in FROM the bank, in which case I suppose you'd want to accommodate inflow along that edge...). On the other hand, if there's a steep downhill slope on one or more sides, then it might make sense to keep that side as open boundaries, and to allow for at least one layer of grid nodes that represent the height of points just downslope of the road, which catch drainage. That seems like a case where you could assume any water reaching such a point will continue to move downslope and away from the road, so you can safely treat it as an "exit boundary". Would it make sense to have different boundary conditions on different sides of the road?

**Comment by amanaster2** [comment]
On both sides of the road, if the water reaches the side of the road, the water should continue to be directed off (on one side, the slope leads to a road-adjacent ditch; the other side is a hillslope leading down)
Example x-sec (not to scale; please ignore the size of the trees 😄):
![image](https://github.com/landlab/landlab/assets/26778806/ca7bd26b-d698-4365-ac92-0d70af958de0)

I think it would make the most sense if my boundary conditions were:
Top: closed
Right: open
Bottom: open
Left: open

I need to ensure that I'm defining these correctly, which seems to be my main problem (and the problem is probably due to me having not used Landlab in quite a while).

**Comment by gregtucker** [comment]
Sounds like you could just set all -9999 nodes to open (BC_NODE_IS_FIXED_VALUE) and the ones where grid.y_of_node >= 5080 to BC_NODE_IS_CLOSED, like

```
grid.status_at_node[elev == -9999] = grid.BC_NODE_IS_FIXED_VALUE
grid.status_at_node[grid.y_of_node >= 5080] = grid.BC_NODE_IS_CLOSED
```

would that work?

**Comment by amanaster2** [comment]
It's *almost* there. I'm still getting drainage areas for the fixed value nodes, which I would expect to be 0. Or am I thinking about this incorrectly?

**Comment by gregtucker** [comment]
The `FlowAccumulator` actually will assign a value of drainage area to a fixed-value node when a core node drains to it, so these boundary nodes won't always have zero drainage area.

**Comment by amanaster2** [comment]
I ended up figuring this out a while ago... And I forgot to close the issue :-)
